### DISCUSSION TITLE
Pipeline

### DISCUSSION THREAD

**Discussion by homesenola** [mtconnect/cppagent#387] [Q&A] [answered]
Hii,

What are the steps that takes place between the data that has come from the adapter(streambuffer) and before it goes to the circularbuffer(boost)
Can Anyone please tell me which part of the code connects to the above context in the pipeline.

Thank you!

**Comment by wsobel** [comment, marked as answer]
The data flow is as follows:

- Data is posted to the pipeline after a block of data is read from the incoming stream.
  - The Start transforms operator() is then first step. 
  - A block being defined as a line or a multiline collection of data
  - The pipelines are described here: https://mtcup.org/Pipeline_Architecture
- The end of the pipeline calls the delivery method in the pipeline context with the type of entity. 
  - The delivery method calls the agent delivery method to place the data in the circular buffer and then forward on to each sink. 

That's about it. Check the code in the pipeline subdirectory for various types of transforms. There is also extensive unit testing for each transform. The transforms and the pipelines are designed to be tested in isolation. 

Best,
Will Sobel
